step-templates/onepassword-retrieve-secrets.json

```diff
@@ -0,0 +1,65 @@
+  "ActionType": "Octopus.Script",
+  "Version": 1,
+  "CommunityActionTemplateId": null,
+  "Packages": [],
+  "GitDependencies": [],
+  "Properties": {
+    "Octopus.Action.Script.ScriptSource": "Inline",
+    "Octopus.Action.Script.Syntax": "PowerShell",
+    "Octopus.Action.Script.ScriptBody": "# Check if the 1Password CLI is installed\nif (-not (Get-Command op -ErrorAction SilentlyContinue)) {\n    Write-Output \"Error: 1Password CLI (op) is not installed.\"\n    exit 1\n}\n\n# Retrieve environment variables from Octopus\n$OP_CONNECT_HOST = (get_octopusvariable \"OnePass.CONNECT_HOST\")\n$OP_CONNECT_TOKEN = (get_octopusvariable \"OnePass.CONNECT_TOKEN\")\n\n# Perform nslookup after removing \"http://\" or \"https://\"\n$hostLookup = $OP_CONNECT_HOST -replace 'https?://', ''\nnslookup $hostLookup\n\n$STEP_NAME = (get_octopusvariable \"Octopus.Step.Name\")\n\n# Retrieve the list of secrets to process from Octopus variable\n$SECRETS = (get_octopusvariable \"OnePass.SecretsManager.RetrieveSecrets.*\")\nWrite-Output $SECRETS\n\n# Validation\nif ([string]::IsNullOrEmpty($SECRETS)) {\n    Write-Output \"Required parameter 'OnePass.SecretsManager.RetrieveSecrets.SecretNames' not specified. Exiting...\"\n    exit 1\n}\n\n# Helper function to save Octopus variable\nfunction Save-OctopusVariable {\n    param (\n        [string]$name,\n        [string]$value\n    )\n\n    set_octopusvariable $name $value --sensitive\n    Write-Output \"Created output variable: ##{Octopus.Action[$STEP_NAME].Output.$name}\"\n}\n\n# Process each secret entry\n$SECRETS -split \"`n\" | ForEach-Object {\n    $secret_entry = $_.Trim()\n    if ([string]::IsNullOrEmpty($secret_entry)) { return }\n\n    # Parse the secret entry\n    $secret_path = ($secret_entry -split '\\|')[0].Trim()  # 1Password path\n    $octopus_variable_name = ($secret_entry -split '\\|')[1].Trim()  # Octopus variable name\n\n    Write-Output \"Fetching secret for path: $secret_path\"\n\n    # Retrieve the secret field using 1Password CLI\n    $field_value = (& op read $secret_path 2>$null)\n\n    # Validate retrieval\n    if ($LASTEXITCODE -ne 0 -or [string]::IsNullOrEmpty($field_value)) {\n        Write-Output \"Error: Failed to retrieve secret for path '$secret_path'.\"\n        exit 1\n    }\n\n    # Save the retrieved value in the specified Octopus variable\n    Save-OctopusVariable -name $octopus_variable_name -value $field_value\n    Write-Output \"Secret retrieval and variable setting complete.\"\n}"
```

Comment:
Does Write-Output $SECRETS just output the names of the secrets?

Reply:
Never mind, I see it's assigned from the parameters for secret names. Looks good to me

Reply:
Is it intentional to have a wildcard in the variable reference when assigning $SECRETS? I'm not sure how get_octopusvariable behaves if there is more than one matching variable:

$SECRETS = (get_octopusvariable \"OnePass.SecretsManager.RetrieveSecrets.*\")

This would probably pick up both SecretNames as expected and the boolean PrintVariableNames, but I can't see any use of that boolean value in the script

Reply:
Hello,  I had written this in a bash script prior to migrating it to powershell. I have updated the script body now to use `OctopusParameters`